anvil/sample.py

```diff
@@ -25,6 +27,175 @@ def sample(loaded_model, action, target_length: int, n_large=20000) -> torch.Ten
+
+    state, log_ratio = sample_batch(
+            loaded_model, action, t_therm, first=True)[2:]
+
+    print(f"Thermalisation: discarded {t_therm} configurations.")
+
+    return state, log_ratio
+
+
+def chain_autocorrelation(loaded_model, action, state, log_ratio) -> float:
+    r"""
+    Compute an observable-independent measure of the integrated autocorrelation
+    time for the Markov chain.
+
+        \tau_int = 0.5 + sum_{\tau=1}^{\tau_max} \rho(\tau)/\rho(0)
+
+    where \rho(\tau)/\rho(0) is the probability of \tau consecutive rejections,
+    which we estimate by
+
+        \rho(\tau)/\rho(0) = # consecutive runs of \tau rejections / (N - \tau)
+
+    See eqs. (16) and (19) in https://arxiv.org/pdf/1904.12072.pdf
+
+    This measure of autocorrelation is used to provide a first guess for an
+    appropriate subsampling interval,
+
+        tskip = ceil(2 * integrated_autocorrelation)
+
+    with the intended effect being that observables on the subsampled chain
+    are entirely decorrelated.
+
+    See http://luscher.web.cern.ch/luscher/lectures/LesHouches09.pdf section 2.2.4
+
+    Parameters
+    ----------
+    loaded_model: Module
+        loaded_model which is going to be used to generate sample states
+    action: Module
+        the action upon which the loaded_model was trained, used to calculate the
+        acceptance condition
+
+    Returns
+    -------
+    state: torch.Tensor
+        the current phi state, for continuity between batches
+    log_ratio: float
+        the ratio associated with state
+    tskip: float
+        Guess for subsampling interval, based on the integrated autocorrelation time
+
+    """
+    batch_size = 1000 # Hard coded num states for estimating integrated autocorrelation
+
+    # Sample some states
+    history, state, log_ratio = sample_batch(
+            loaded_model, action, batch_size, state, log_ratio
+    )[1:]
+
+    N = len(history)
+    autocorrelations = torch.zeros(N - 1, dtype=torch.float)
+    consecutive_rejections = 0
+
+    for step in history:
+        if step == True:  # move accepted
+            if consecutive_rejections > 0:  # faster than unnecessarily accessing array
+                autocorrelations[1 : consecutive_rejections + 1] += 1
+            consecutive_rejections = 0
+        else:  # move rejected
+            consecutive_rejections += 1
+    if consecutive_rejections > 0:  # pick up last rejection run
+        autocorrelations[1 : consecutive_rejections + 1] += 1
+
+    # Compute integrated autocorrelation
+    integrated_autocorrelation = 0.5 + torch.sum(
+        autocorrelations / (N - torch.arange(N - 1))
+    )
+
+    tskip = ceil(2 * integrated_autocorrelation)
+    print(
+        f"Guess for sampling interval: {tskip}, based on {batch_size} configurations."
+    )
+
+    return state, log_ratio, tskip
+
+
+def sample(loaded_model, action, target_length: int) -> torch.Tensor:
```

Comment:
ok so this is more of a reportengine thing:
rename `thermalisation` -> `thermalised_state` and have it just return a single state

then make the signature of `chain_autocorrelation`: 

```python
def chain_autocorrelation(loaded_model, action, thermalised_state):
```

then signature of `sample`:

```python
def sample(loaded_model, action, target_length: int, thermalised_state, chain_autocorrelation)
```
and you can skip the first bit of the function because report engine will automatically evaluate `thermalised_state` and `chain_autocorrelation`

the nice part of this is that both of those functions will also be valid `actions` and so if I wanted to I could calculate auto correlation without having to do a full sample

Reply:
that's rather cool

Reply:
I couldn't get it to work though. I always seem to run into an error if I try to use these reportengine actions as normal functions to which I can give arguments which aren't other actions.

Serious question: am I being incredibly thick?

Reply:
urm can you approximately write down signatures of functions that you tried

so just the `def function(arg1, arg2)` etc. and their actual names for relevant functions

Reply:
for me this works:

```python
def thermalised_state(loaded_model, action):
    ...
    return states[-1]


def chain_autocorrelation(loaded_model, action, thermalised_state) -> float:
    ...
    # Sample some states
    states, history = sample_batch(
            loaded_model, action, batch_size, thermalised_state
    )
    ...
    integrated_autocorrelation = 0.5 + torch.sum(
        autocorrelations / (N - torch.arange(N - 1, dtype=torch.float))
    ) # here you had a mismatch of types
    ...
    return states[-1], sample_interval


def sample(loaded_model, action, target_length: int, thermalised_state, chain_autocorrelation) -> torch.Tensor:
    ...
    current_state = thermalised_state
    _, sample_interval = chain_autocorrelation
    ...
    return decorrelated_chain
```

Reply:
since chain_autocorrelation is also returning a final state I could even remove thermalised state from sample and just do `current_state, sample_interval = chain_autocorrelation`

however I really would be tempted to make chain_autocorrelation return just a number - we don't really care about thermalising further we just want to know the autocorrelation

Reply:
ah that's grand. My mistake was to try to do `current_state = thermalised_state(loaded_model, action)` etc.